From 65e852d74a296e1e8ca93c20cb72164bae064634 Mon Sep 17 00:00:00 2001
From: Phil Edworthy <phil.edworthy@renesas.com>
Date: Wed, 27 Apr 2016 08:39:44 +0100
Subject: [PATCH] dma: dw: Optimise memory width when peripheral is the flow
 controller

If the peripheral is the flow controller, we can set the width of
accesses to memory irrespective of the data length. This improves
the bus utilisation.

Signed-off-by: Phil Edworthy <phil.edworthy@renesas.com>
---
 drivers/dma/dw/core.c | 12 ++++++++++--
 1 file changed, 10 insertions(+), 2 deletions(-)

diff --git a/drivers/dma/dw/core.c b/drivers/dma/dw/core.c
index e1b3ed9..fced1da 100644
--- a/drivers/dma/dw/core.c
+++ b/drivers/dma/dw/core.c
@@ -780,7 +780,11 @@ dwc_prep_slave_sg(struct dma_chan *chan, struct scatterlist *sgl,
 			mem = sg_dma_address(sg);
 			len = sg_dma_len(sg);
 
-			mem_width = __ffs(data_width | mem | len);
+			if (!sconfig->device_fc) {
+				mem_width = __ffs(data_width | mem | len);
+			} else {
+				mem_width = __ffs(data_width | mem);
+			}
 
 slave_sg_todev_fill_desc:
 			desc = dwc_desc_get(dwc);
@@ -834,7 +838,11 @@ dwc_prep_slave_sg(struct dma_chan *chan, struct scatterlist *sgl,
 			mem = sg_dma_address(sg);
 			len = sg_dma_len(sg);
 
-			mem_width = __ffs(data_width | mem | len);
+			if (!sconfig->device_fc) {
+				mem_width = __ffs(data_width | mem | len);
+			} else {
+				mem_width = __ffs(data_width | mem);
+			}
 
 slave_sg_fromdev_fill_desc:
 			desc = dwc_desc_get(dwc);
-- 
2.7.4

